MarkBind/markbind
# Adopt TypeScript for core package

**What is the purpose of this pull request?**

- [x] Others, please explain: Project development upgrade

First phase of #1421

**Overview of changes:**
(Edited as of 17/04)
Some of the `.js` code files in `core` have been renamed as `.ts` files and adapted into TypeScript as a starter.

See the documentation on migrating to TypeScript: https://deploy-preview-1877--markbind-master.netlify.app/devguide/migratingtotypescript

Workflow of `core` and by extension the whole project are slightly changed: 
- Development 
  - Three new scripts are added in root, `npm run build:backend` to build the TypeScript files, `npm run clean` to remove built files, and `npm run dev` to start the development environment which builds the files and automatically rebuilds on changes
  - IDEs like WebStorm, VSCode, natively supports rebuild on change too, so developers who use those IDEs can sidestep `npm run dev`.
  - Three new scripts are added in core, `npm run build` to build the TypeScript files, `npm run clean` to remove built files, and `npm run prepare` which is explained in Installation and Publishing below
- Installation
  - A new lifecycle script `prepare` is added in core which does `npm run build` and automatically runs as part of installation of the package (`npm run setup`). 
- Debugging
  - Builds will emit sourcemaps (`.js.map`) which provides the ability to directly debug `.ts` files (e.g. put breakpoints, etc)
- Linting
  - New package `typescript-eslint` is integrated to `eslint` to enable lint on `.ts` files
  - `npm run lint` now lints `.ts` files
- Testing
  - New package `ts-jest` is integrated to `jest` to perform direct unit tests to `.ts` files without the need to manually build them
  - However, functional tests with `cli` will require for the files to be built first
- Publishing
  - Builds will emit declaration files (`.d.ts`) and their maps (`.d.ts.map`) for type-checking and definitions support across projects
  - The lifecycle `prepare` script in core will automatically run as part of publishing of the package (`lerna publish`) 

**Anything you'd like to highlight / discuss:**
- Comments on the approach taken are appreciated.

**Testing instructions:**
Setup test: `npm run setup` should compile the TypeScript files.
Development test: `npm run test` should be fine.
Development usage test (with linked `markbind-cli`): Any `markbind` commands should be fine.

**Proposed commit message: (wrap lines at 72 characters)**
Adopt TypeScript for core package

The dynamic typing nature of JavaScript can create avenues for subtle
bugs during runtime related to parameter types. Strict typing system,
like TypeScript, can address such cases and benefit development of
MarkBind.

Let's introduce TypeScript to the core package to improve the quality
and security of our codebase. Furthermore, let's adapt starter files
to TypeScript as an example, and document on the steps to migrate other
files to TypeScript for developers.

---

**Checklist:** :ballot_box_with_check:
